---
title: "Report from mass spectrometry data analysis pipeline"
auhtor: "`r sessionInfo()$platform`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
   rmd: "report.Rmd"
output:
  html_document:
    theme: cosmo
    toc: yes
    toc_depth: 2
    number_sections: yes
    df_print: paged
---

----------

# Background

The pipeline is built using [snakemake](https://snakemake.readthedocs.io/en/stable/) and processes MS data using the following steps:

1. Prepare `workflow` file (`python` script)
2. Fetch protein database from NCBI or use user-supplied fasta file (`python`, [NCBI Datasets](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/))
3. Generate decoy proteins ([DecoyPyrat](https://github.com/wtsi-proteomics/DecoyPYrat))
4. Import raw files, search protein database ([fragpipe](https://fragpipe.nesvilab.org/))
5. Align feature maps using IonQuant ([fragpipe](https://fragpipe.nesvilab.org/))
6. Import quantified features, infer and quantify proteins ([R MSstats](https://www.bioconductor.org/packages/release/bioc/html/MSstats.html))
7. Compare different biological conditions, export results ([R MSstats](https://www.bioconductor.org/packages/release/bioc/html/MSstats.html))
8. Generate HTML report with embedded QC plots ([R markdown](https://rmarkdown.rstudio.com/))
9. Generate PDF report from HTML [weasyprint](https://weasyprint.org/)
10. Clean up temporary files after pipeline execution (`bash` script)


----------

# Prerequisites

## Packages

Loaded required R packages:

- `tidyverse`
- `ggrepel`
- `scales`
- `dendextend`
- `ggpubr`


```{r, echo = FALSE, warning = FALSE}
suppressPackageStartupMessages({
  library(tidyverse)
  library(ggrepel)
  library(scales)
  library(dendextend)
  library(ggpubr)
})
```


```{r, echo = FALSE}
# custom ggplot2 theme that is reused for all later plots
custom_colors <- c("#E7298A", "#66A61E", "#E6AB02", "#7570B3", "#B3B3B3", "#1B9E77", "#D95F02", "#A6761D")
custom_range <- function(n = 5) {
  colorRampPalette(custom_colors[c(1, 5, 2)])(n)
}

custom_theme <- function(base_size = 12, base_line_size = 1.0, base_rect_size = 1.0, ...) {
  theme_light(base_size = base_size, base_line_size = base_line_size, base_rect_size = base_rect_size) + theme(
    title = element_text(colour = grey(0.4), size = 10),
    plot.margin = unit(c(12, 12, 12, 12), "points"),
    axis.ticks.length = unit(0.2, "cm"),
    axis.ticks = element_line(colour = grey(0.4), linetype = "solid", lineend = "round"),
    axis.text.x = element_text(colour = grey(0.4), size = 10),
    axis.text.y = element_text(colour = grey(0.4), size = 10),
    panel.grid.major = element_line(linewidth = 0.6, linetype = "solid", colour = grey(0.9)),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(linetype = "solid", colour = grey(0.4), fill = NA, linewidth = 1.0),
    panel.background = element_blank(),
    strip.background = element_blank(),
    strip.text = element_text(colour = grey(0.4), size = 10, margin = unit(rep(3, 4), "points")),
    legend.text = element_text(colour = grey(0.4), size = 10),
    legend.title = element_blank(),
    legend.background = element_blank(),
    ...
  )
}

# set graphical parameter for subfigure labels
list_fontpars <- list(face = "plain", size = 14)

# function to export an image as svg and png
save_plot <- function(pl, path = "../figures/", width = 6, height = 6) {
  pl_name <- deparse(substitute(pl))
  svg(
    filename = paste0(path, pl_name, ".svg"),
    width = width, height = height
  )
  print(pl)
  dev.off()
  png(
    filename = paste0(path, pl_name, ".png"),
    width = width * 125, height = height * 125, res = 120
  )
  print(pl)
  invisible(capture.output(dev.off()))
}

# path to export figures
path_figures <- str_remove(snakemake@output[[1]], "report\\.html$")
```


## Result tables

Output tables from `MSstats` that were imported:

*NOTE: tables are only rendered in HTML output but not PDF.*

```{r, echo = FALSE}
df_tables <- snakemake@input[6:9] %>%
  unlist() %>%
  enframe(name = "table", value = "path")

for (row in 1:nrow(df_tables)) {
  assign(
    x = paste0("df_", df_tables[[row, "table"]]),
    value = read_csv(df_tables[[row, "path"]], show_col_types = FALSE)
  )
}

df_tables
```

## Sample sheet

Sample sheet listing all conditions (experimental groups) and replicates:

*NOTE: tables are only rendered in HTML output but not PDF.*

```{r, echo = FALSE}
df_sample_sheet <- read_tsv(
  snakemake@config$samplesheet,
  show_col_types = FALSE,
  col_names = c("file_path", "condition", "replicate", "ms_type", "control")
)
mutate(df_sample_sheet, file_path = paste0("..", str_sub(file_path, -35L, -1L)))
```

```{r, echo = FALSE}
# code chunk to set figure proportions based on N samples
figwidth <- 8.5
figheight <- 1 + round(nrow(df_sample_sheet) / 2.8, 1)
figheight2 <- 1 + 4 * ceiling(length(unique(df_comparison_result$Label)) / 3)
figheight3 <- figheight + 0.6
```

```{r, echo = FALSE}
# code chunk to re-number replicate; MSstats uses consecutive
# numbers ("SUBJECTS") to indicate different sources of a sample;
# re-occuring replicates across conditions mean the same subject was sampled;
# but more common is treating replicate index as independent per condition
df_protein_level_data <- df_protein_level_data %>%
  group_by(GROUP) %>%
  mutate(replicate = as.numeric(as.factor(SUBJECT))) %>%
  ungroup

df_feature_level_data <- df_feature_level_data %>%
  group_by(GROUP) %>%
  mutate(replicate = as.numeric(as.factor(SUBJECT))) %>%
  ungroup
```

----------

# Quality control

## Summary statistics

### Number of quantified peptides

- Quantified peptides are used to calculate protein quantity
- The higher the number of quantified peptides the better
- Peptide number largely depends on the experimental set up

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
plot_quant_pept <- df_feature_level_data %>%
  group_by(GROUP, replicate) %>%
  filter(is.na(predicted)) %>%
  summarize(.groups = "drop", n_proteins = length(unique(FEATURE))) %>%
  ggplot(aes(x = n_proteins, y = paste(GROUP, replicate))) +
  geom_col(color = "white", fill = custom_colors[1]) +
  labs(x = "", y = "", title = "Number of quantified peptides") +
  custom_theme()

# save_plot(plot_quant_pept, path = path_figures)
print(plot_quant_pept)
```

### Number of quantified proteins

- Proteins are quantified from peptide (feature) intensity
- Protein quantification is performed by calculating the sum or mean of feature intensity
- A good coverage for whole cell lysate is > 50% of annotated proteins quantified
- This number can be highly condition-dependent

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
plot_quant_prot <- df_protein_level_data %>%
  group_by(GROUP, replicate) %>%
  summarize(.groups = "drop", n_proteins = length(unique(Protein))) %>%
  ggplot(aes(x = n_proteins, y = paste(GROUP, replicate))) +
  geom_col(color = "white", fill = custom_colors[1]) +
  labs(x = "", y = "", title = "Number of quantified proteins") +
  custom_theme()

# save_plot(plot_quant_prot, path = path_figures)
print(plot_quant_prot)
```

### Number of features (peptides) per protein

- Number of quantified features used to calculate protein quantity
- Bars represent the total number of proteins, color indicates coverage group
- The higher the coverage (*n* features) per protein the better
- Proteins quantified with only one peptide are common and can be acceptable

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight3}
plot_feat_per_prot <- df_protein_level_data %>%
  mutate(
    n_features = cut(NumMeasuredFeature,
      breaks = c(0, 5, 10, 20, 50, Inf)
    )
  ) %>%
  group_by(GROUP, replicate) %>%
  count(n_features) %>%
  ggplot(aes(x = n, y = paste(GROUP, replicate), fill = n_features)) +
  geom_col(color = "white") +
  labs(x = "", y = "", title = "Number of features (peptides) per protein") +
  custom_theme(legend.position = "bottom") +
  scale_fill_manual(values = custom_colors)

# save_plot(plot_feat_per_prot, path = path_figures)
print(plot_feat_per_prot)
```


## Missing values

### Missing values on feature level

- `MSstats` attempts to infer missing feature intensities based on features quantified in other samples
- The lower the number of predicted or inferred features, the better
- Missing features can not always be inferred
- A higher number of replicates can significantly reduce missing features

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight3}
plot_missing_1 <- df_feature_level_data %>%
  select(PROTEIN, GROUP, replicate, predicted) %>%
  group_by(GROUP, replicate) %>%
  summarize(
    .groups = "drop",
    measured = sum(is.na(predicted)),
    predicted = sum(!is.na(predicted))
  ) %>%
  pivot_longer(c("measured", "predicted"),
    names_to = "feature", values_to = "count"
  ) %>%
  mutate(feature = factor(feature, c("predicted", "measured"))) %>%
  ggplot(aes(x = count, y = paste(GROUP, replicate), fill = feature)) +
  geom_col(color = "white", position = "stack") +
  labs(x = "", y = "", title = "Number of predicted and measured features") +
  custom_theme(legend.position = "bottom") +
  scale_fill_manual(values = custom_colors[c(2, 1)])

# save_plot(plot_missing_1, path = path_figures)
print(plot_missing_1)
```

### Missing values on protein level

- This plot shows percent (%) feature inference that was used to quantify each individual protein
- None-imputed (fully quantified) proteins were removed to make the figure appear more clean
- The more skewed the distribution is to the left side (0 %), the better
- The majority of proteins should have lower than 25% imputed features

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
plot_missing_2 <- df_protein_level_data %>%
  select(Protein, GROUP, replicate, MissingPercentage) %>%
  filter(MissingPercentage > 0) %>%
  ggplot(aes(x = MissingPercentage * 100, y = paste(GROUP, replicate))) +
  geom_violin(color = "white", fill = custom_colors[1], alpha = 0.3) +
  geom_jitter(height = 0.2, color = custom_colors[1]) +
  labs(
    x = "", y = "", title = "Percent imputed features per protein"
  ) +
  custom_theme()

# save_plot(plot_missing_2, path = path_figures)
print(plot_missing_2)
```


### Problematic protein quantifications

- This plot shows protein quantifications where more than 50% of features are missing
- These proteins are flagged by `MSstats` 'as problematic'
- The proportion of problematic quantifications should not exceed 10% of all protein quantifications

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight}
plot_problematic <- df_protein_level_data %>%
  group_by(GROUP, replicate) %>%
  count(more50missing) %>%
  filter(more50missing) %>%
  ggplot(aes(x = n, y = paste(GROUP, replicate))) +
  geom_col(color = "white", fill = custom_colors[1]) +
  labs(x = "", y = "", title = "Number of proteins with > 50% missing features") +
  custom_theme()

# save_plot(plot_problematic, path = path_figures)
print(plot_problematic)
```

### Number of proteins quantified in number of samples

- Step plot showing number (N) of proteins quantified in number of samples
- Large area under the curve means that many proteins have been quantified in a majority of the samples
- Small area under the curve (early descending steps) means that few proteins have been quantified in majority of samples

```{r, echo = FALSE, warning = FALSE, fig.width = 7, fig.height = 7}
plot_stepplot <- df_protein_level_data %>%
  group_by(Protein) %>%
  summarize(n_samples = sum(!is.na(LogIntensities))) %>%
  count(n_samples, name = "n_proteins") %>%
  arrange(desc(n_samples)) %>%
  add_row(n_samples = nrow(df_sample_sheet), n_proteins = 0, .before = 0) %>%
  add_row(n_samples = 0, n_proteins = 0) %>%
  mutate(n_proteins = cumsum(n_proteins)) %>%
  ggplot(aes(x = n_proteins, y = n_samples)) +
  geom_step(color = custom_colors[1], linewidth = 1, direction = "vh") +
  labs(
    x = "N proteins", y = "N samples",
    title = "N proteins quantified in N samples"
  ) +
  custom_theme()

# save_plot(plot_stepplot, path = path_figures)
print(plot_stepplot)
```

----------

# Protein quantification

## Sample-wise comparison

### Normalized log2 protein abundance

- Protein abundance calculated by MSstats after normalization
- The default is per-group median normalization
- Normalization options can be changed in the `config.yml` file supplied with this pipeline
- Color indicates the number of the individual replicates
- Similarity between replicates should be higher than similarity between conditions

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight3}
plot_log2_abundance <- df_protein_level_data %>%
  group_by(GROUP, Protein) %>%
  arrange(replicate) %>%
  mutate(replicate = seq_along(LogIntensities)) %>%
  ggplot(aes(x = LogIntensities, y = GROUP, color = factor(replicate))) +
  geom_vline(xintercept = 0, color = grey(0.75), linetype = 2) +
  geom_boxplot(outlier.shape = NA) +
  labs(x = "", y = "", title = "Normalized log2 protein abundance") +
  custom_theme(legend.position = "bottom") +
  scale_color_manual(values = colorRampPalette(custom_colors)(
    length(unique(df_protein_level_data$replicate))
  ))

# save_plot(plot_log2_abundance, path = path_figures)
print(plot_log2_abundance)
```

### Profile plots

- Comparison of feature intensity profile for selected protein(s)
- This plot can be used to evaluate the protein quantification from feature intensity
- This is a `ggplot` adaption of the default profile plot output from `MSstats`
- A selection of 10 proteins with different total abundance is chosen as representative sample
- Each protein represents the middle value of one of 10 abundance ranges from low (top panel) to high (bottom)
- Good quantification means that grey lines (features) go in parallel
- Bad quantification means that there are broken lines (missing quant.) and sudden spikes (high variablity) 

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = 12}
list_diff_abund_prot <- df_protein_level_data %>%
  group_by(Protein) %>%
  summarize(median_int = median(LogIntensities, na.rm = TRUE)) %>%
  mutate(int_group = cut_interval(median_int, n = 10)) %>%
  group_by(int_group) %>%
  arrange(median_int) %>%
  slice(round(n() / 2)) %>%
  pull(Protein)

plot_profile <- ggplot() +
  geom_line(
    data = df_feature_level_data %>%
      filter(PROTEIN %in% list_diff_abund_prot),
    mapping = aes(x = factor(RUN), y = ABUNDANCE, group = PEPTIDE),
    color = grey(0.8)
  ) +
  geom_point(
    data = df_protein_level_data %>%
      rename(PROTEIN = Protein) %>%
      filter(PROTEIN %in% list_diff_abund_prot),
    mapping = aes(x = factor(RUN), y = LogIntensities, color = GROUP),
    size = 2
  ) +
  labs(
    x = "", y = expression("log"[2] * " intensity"),
    title = "Feature intensity (grey) and inferred protein quantity (color)"
  ) +
  custom_theme(legend.position = "bottom") +
  scale_color_manual(values = colorRampPalette(custom_colors)(
    length(unique(df_protein_level_data$GROUP))
  )) +
  facet_grid(rows = vars(factor(PROTEIN, list_diff_abund_prot)))

# save_plot(plot_profile, path = path_figures, width = 8, height = 12)
print(plot_profile)
```

### Similarity of conditions and replicates

- Principal component analysis (PCA) to visualize similarity between samples
- Conditions are encoded by color, and replicates are indicated with numbers

```{r, echo = FALSE, fig.width = 7, fig.height = 7, warning = FALSE}
pca_result <- df_protein_level_data %>%
  select(Protein, GROUP, replicate, LogIntensities) %>%
  pivot_wider(names_from = c("GROUP", "replicate"), values_from = "LogIntensities") %>%
  column_to_rownames("Protein") %>%
  as.matrix() %>%
  t() %>%
  replace(., is.na(.), 0) %>%
  prcomp()

df_PCA <- pca_result$x %>%
  as_tibble(rownames = "sample") %>%
  mutate(
    condition = str_sub(sample, 1, -3),
    replicate = str_sub(sample, -1, -1)
  )

if (any(nchar(df_PCA$condition) > 25)) {
  df_PCA <- df_PCA %>%
    mutate(condition = paste0(
      "(", as.numeric(as.factor(condition)), ") ", substr(condition, 1, 25), "..."
    ))
}

plot_pca <- df_PCA %>%
  ggplot(aes(
    x = PC1, y = -PC2, size = PC3,
    color = condition, label = replicate
  )) +
  geom_point(alpha = 0.7) +
  geom_text_repel(size = 4, show.legend = FALSE) +
  labs(
    title = "PCA, first three principal components",
    subtitle = "Point size encodes PC3", x = "PC1", y = "PC2"
  ) +
  custom_theme(legend.position = "bottom", aspect = 1) +
  scale_color_manual(values = colorRampPalette(custom_colors)(
    length(unique(df_protein_level_data$GROUP))
  )) +
  guides(size = "none") +
  guides(color = guide_legend(ncol = 3))

# save_plot(plot_pca, path = path_figures, width = 7, height = 7)
print(plot_pca)
```

## Group-wise comparison

### Model QC

Internally, MSstats fits statistical models to protein quantifications for each comparison, similar to mixed linear models.
Such models can be fitted for multiple comparisons between conditions, particularly when multi-factorial data sets should be analyzed.
This entails for example conditions characterized by different variables simultaneously, such as `minimal medium` and `stationary phase`
versus `rich medium` and `logarithmic phase` versus `...` etc.

To check if the assumptions of the model fitting are met, a [QQ (quantile quantile) plot](https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot) 
can be used. The most important assumption is the normal distribution of measurements and errors. Errors should scale with the measurement,
and QQ plots test this assumption by plotting the theoretical quantiles against the observed quantiles of the measurements. Briefly:

- QQ plots show the expected versus experimentally observed variation
- The ideal behavior is that the observations follow a line
- If not, protein quantifications are not normal-distributed
- The following figure shows QQ plots for the 10 exemplary proteins from the previous section [#profile_plots](Profile plots).

```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = 12}
plot_qq <- df_model_qc %>%
  filter(Protein %in% list_diff_abund_prot) %>%
  ggplot(aes(sample = ABUNDANCE, color = GROUP, group = NA)) +
  stat_qq(color = grey(0.5)) +
  stat_qq_line() +
  labs(
    title = "QQ plot, theoretical vs. observed quantiles of feature abundance",
    x = "theoretical quantiles", y = "variable"
  ) +
  custom_theme(aspect = 1) +
  facet_wrap(~Protein, ncol = 3, nrow = 4, scales = "free")

# save_plot(plot_qq, path = path_figures, width = 9, height = 12)
print(plot_qq)
```

### Volcano plot

```{r, echo = FALSE, warnings = FALSE}
threshold_pvalue <- 0.01
threshold_fc <- 2
```

Volcano plots are useful to evaluate the global changes between conditions. They show the log2 fold change (`log2FC`)
on the x-axis and the p-value on the y-axis. Both metrics are computed by `MSstats` for the comparisons ('constrasts')
indicated in the sample sheet. The p-value is the Benjamini-Hochberg adjusted p-value from `MSstats`' default statistical test.

- Significance boundaries indicate which genes are significant in terms of **p-value** *and* **effect size** (log2 fold change)
- The standard boundaries for both metrics are:
- alpha significance threshold = `r threshold_pvalue`
- fold change threshold = `r threshold_fc`
- The boundary represents the *product* of fold change and p-value thresholds
- Example: All points on this line are equivalent to _k_ = -log10(`r threshold_pvalue`) * `r threshold_fc` = `r -log10(threshold_pvalue) * threshold_fc`.


```{r, echo = FALSE, warning = FALSE, fig.width = figwidth, fig.height = figheight2}
geom_effect_boundaries <- function(alpha, fc, xlim = c(-8, 8), color = grey(0.3)) {
  x <- c(seq(xlim[1], 0, 0.1), seq(0, xlim[2], 0.1))
  y <- (-log10(alpha) * log2(fc)) / c(seq(xlim[2], 0, -0.1), seq(0, xlim[2], 0.1))
  df <- data.frame(x = x, y = y)
  geom_line(data = df, aes(
    x = x, y = y,
    shape = NULL, col = NULL
  ), lty = 2, color = color)
}

plot_volcano <- df_comparison_result %>%
  ggplot(aes(
    x = log2FC, y = -log10(adj.pvalue),
    color = abs(log2FC) * -log10(adj.pvalue) >=
      log2(threshold_fc) * -log10(threshold_pvalue)
  )) +
  geom_point(size = 0.5) +
  geom_effect_boundaries(
    alpha = threshold_pvalue,
    fc = threshold_fc,
    xlim = c(-15, 15)
  ) +
  labs(
    title = "Volcano plot",
    subtitle = paste0(
      "color indicates proteins with score equivalent to log2FC > ",
      threshold_fc, " and p-value < ", threshold_pvalue
    ),
    x = expression("log"[2] * " FC"), y = expression("-log"[10] * " p-value")
  ) +
  lims(x = c(-7, 7), y = c(-0.1, 8.1)) +
  custom_theme(legend.position = "bottom", aspect = 1) +
  scale_color_manual(values = custom_colors[c(5, 1)]) +
  facet_wrap(~Label,
    ncol = 3,
    nrow = ceiling(length(unique(df_comparison_result$Label)) / 3)
  )

# save_plot(plot_volcano, path = path_figures)
print(plot_volcano)
```

### List of differentially abundant proteins

- The table lists all proteins that have been marked with colored points in the volcano plot
- The table is in long format: Quantifications are shown in one column and metadata others

*NOTE: tables are only rendered in HTML output but not PDF.*

```{r, echo = FALSE, warnings = FALSE}
df_significant <- df_comparison_result %>%
  mutate(score = abs(log2FC) * -log10(adj.pvalue)) %>%
  group_by(Protein) %>%
  filter(any(
    !is.na(pvalue) &
      (score >= (log2(threshold_fc) * -log10(threshold_pvalue)))
  )) %>%
  ungroup()

if (nrow(df_significant)) {
  print(paste0(
    "A total of ", length(unique(df_significant$Protein)),
    " proteins were found with at least one significant change in a comparison"
  ))
  df_significant
} else {
  print("No significantly changed proteins have been found with the default thresholds.")
}
```

### Heat map

- The heat map shows a direct comparison of protein abundance for all conditions
- Included are only the proteins falling below the previously defined thresholds

```{r, echo = FALSE, warnings = FALSE}
# originally contained in R package https://github.com/m-jahn/R-tools
fct_cluster <- function(variable, group, value, method = "ward.D2") {
  variable <- as.character(variable)
  df <- tidyr::tibble(variable = variable, group = group, value = value)
  df <- tidyr::complete(df, variable, group)
  df <- tidyr::pivot_wider(df,
    names_from = group, values_from = value,
    values_fn = function(x) {
      replace(x, is.infinite(x), NA)
    }
  )
  mat <- as.matrix(tibble::column_to_rownames(df, var = "variable"))
  dist_mat <- replace(dist(mat), is.na(dist(mat)), 0)
  cl <- stats::hclust(dist_mat, method = "ward.D2")
  ord <- stats::order.dendrogram(stats::as.dendrogram(cl))
  factor(variable, rownames(mat)[ord])
}
```


```{r, echo = FALSE, warnings = FALSE, fig.width = 6, fig.height = 8}
if (nrow(df_significant)) {
  # reduce number of proteins before plotting heat map
  if (length(unique(df_significant$Protein)) > 50) {
    df_significant <- df_significant %>%
      group_by(Label) %>%
      arrange(desc(abs(log2FC))) %>%
      slice(1:50)
    subtitle_reduced <- "Number of proteins was reduced to 50."
  } else {
    subtitle_reduced <- ""
  }
  plot_heatmap <- df_significant %>%
    mutate(Protein = fct_cluster(Protein, Label, log2FC)) %>%
    ggplot(aes(
      y = Protein,
      x = Label, fill = log2FC
    )) +
    geom_tile() +
    custom_theme(legend.pos = "right") +
    labs(
      title = paste0("Heat map of protein abundance (n = ", length(unique(df_significant$Protein)), ")"),
      subtitle = subtitle_reduced,
      x = "",
      y = ""
    ) +
    theme(
      axis.text.x = element_text(angle = 90),
      axis.ticks.x = element_blank()
    ) +
    scale_fill_gradientn(
      colours = c(custom_colors[1], grey(0.9), custom_colors[2]),
      limits = c(-6, 6)
    )

  # save_plot(plot_heatmap, path = path_figures)
  print(plot_heatmap)
} else {
  print("No significantly changed proteins have been found with the default thresholds.")
}
```

----------

# About this report

## Pipeline

This report was automatically generated by the [snakemake-ms-proteomics](https://github.com/MPUSP/snakemake-ms-proteomics) pipeline.

For issues, bugs, and feature requests please use the pipeline's [github page](https://github.com/MPUSP/snakemake-ms-proteomics/issues).

For all other feedback, contact the author(s):

- Michael Jahn, PhD (author and maintainer): jahn@mpusp.mpg.de

The pipeline is developed at the [Max Planck Unit for the Science of Pathogens](https://www.mpusp.mpg.de/), Berlin, Germany.

## Packages used in the pipeline

This list shows the fixed dependencies used as `conda` environment variables within the pipeline.

```{r, echo = FALSE, warnings = FALSE}
versions <- read_delim(snakemake@input$versions,
  delim = "\n",
  show_col_types = FALSE,
  col_names = "module"
)

print(versions, n = Inf)
```

## Data accessability

The following resources were used to generate this report:

- MSstats result tables: `r paste0(snakemake@config$output, "msstats/")`.
- sample sheet: `r snakemake@config$samplesheet`
- fragpipe workflow: `r snakemake@config$workflow`
- organism database: `r snakemake@config$database`
- fragpipe binary: `r snakemake@config$fragpipe$path`

## Session Info

Link to the R markdown source that was used to generate this report:

<a download="report.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">R Markdown source file</a>

Session info (R base and package versions):

```{r, echo = FALSE}
sessionInfo()
```

